File: insertion_sort.rb
def insertion_sort(nums)
  return unless nums.instance_of?(Array)

  nums.each_with_index do |_num, i_start|
    j_run = i_start - 1
    while j_run > -1
      swap(nums, j_run + 1, j_run) if nums[j_run + 1] > nums[j_run]
      j_run -= 1
    end
  end
  nums
end

def swap(nums, r, l)
  return unless nums.instance_of?(Array)

  tmp = nums[r]
  nums[r] = nums[l]
  nums[l] = tmp
end

p insertion_sort([1, 2, 3, 4])

File: insertion_sort.py
def insertion_sort(nums):
    for i_start in range(len(nums)):
        for j_run in range(i_start -1 , -1 , -1):
          # range條件，從倒數第二個開始，若判斷j_run = -1 時停止 , 每次都減一 ->往前一個
            if nums[j_run + 1] > nums[j_run]:
                swap(nums, j_run + 1 , j_run)
def swap(nums , r , l):
    tmp = nums[r]
    nums[r] = nums[l]
    nums[l] = tmp
    
if __name__ == "__main__":
    nums = [8, 2 ,6 ,10, 4]
    insertion_sort(nums)
    print(nums)